### PR TITLE
	Respect SET/RESET SESSION queries

### PR DESCRIPTION
**Pull request by saurfang**
According #32, parameters can already be set via `dbConnect(..., parameters = list(...)`.

The Python implementation of Presto API appears to respect [`SET/RESET SESSION ...`](https://prestodb.io/docs/current/sql/set-session.html) query commands directly. This PR does the same for RPResto according to [HTTP Protocol](https://github.com/prestodb/presto/wiki/HTTP-Protocol) specification.

One caveat is this is only possible via `dbGetQuery` due to the API behavior, where parameters are managed by the client instead of the server.

I had to add a new RCpp function to ensure parameters are updated in-place instead of being copied upon modification inside the function.
Furthermore, `PrestoResult` doesn't have access to the connection object and therefore gains a new function slot to do the update.

Test cases are added and tested with a local cluster.

Updated to Roxygen 6.0.1

### CODE REVIEW

File: RPresto.Rproj [hunk 1]
```diff
@@ -5,8 +5,13 @@ SaveWorkspace: No
 AlwaysSaveHistory: Default
 
 EnableCodeIndexing: Yes
+UseSpacesForTab: Yes
```
Comment (onurfiliz) [inline RPresto.Rproj]: Could you remove the RPresto.Rproj changes? I'd prefer to keep the context of the PR limited to session parameters.

File: tests/testthat/test-session.property.R [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+# LICENSE file in the root directory of this source tree. An additional grant
+# of patent rights can be found in the PATENTS file in the same directory.
+
+context("session.property")
+
+source('utilities.R')
+
+query_session_property <- function(conn, property) {
+  properties <- dbGetQuery(conn, 'SHOW SESSION')
+  properties[properties$Name == property, ]$Value
```
Comment (onurfiliz) [inline tests/testthat/test-session.property.R]: Please use list[['item']] style of indexing for accessing list items.

You can also do:

    return(properties[properties[['Name']] == property, 'Value', drop=TRUE])

File: tests/testthat/test-session.property.R [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+}
+
+test_that('session properties can be configured in connection', {
+  conn <- setup_live_connection()
+  expect_equal(conn@parameters, list())
+
+  # verify default value is applied
+  test_property <- 'optimize_hash_generation'
+  expect_equal(query_session_property(conn, test_property), 'true')
```
Comment (onurfiliz) [inline tests/testthat/test-session.property.R]: Even though it's a rather safe assumption, this default value would not be guaranteed. A better option would be to take the default value and invert it in the following tests.

File: R/dbFetch.R [hunk 4]
```diff
@@ -75,7 +99,10 @@ NULL
     rv[[chunk.count]] <- chunk
     chunk.count <- chunk.count + 1
   }
-  if (length(rv) == 1) {
+  if (length(rv) == 0) {
```
Comment (onurfiliz) [inline R/dbFetch.R]: I'd prefer if we didn't do this. If the server does not return anything even when we have fetched all chunks, then it's either a server bug or we are not handling errors properly at some other spot.

File: R/dbFetch.R [hunk 5]
```diff
@@ -61,6 +61,30 @@ NULL
       res@cursor$stats(content[['stats']])
       stop.with.error.message(content)
     }
+
+    # Handle SET/RESET SESSION updates
+    if (!is.null(content[['updateType']])) {
+      switch(
+        content[['updateType']],
+        'SET SESSION' = {
+          properties <- get.response$headers[['x-presto-set-session']]
```
Comment (onurfiliz) [inline R/dbFetch.R]: Can you use `httr::headers()` instead?

File: R/dbSendQuery.R [hunk 6]
```diff
@@ -49,7 +49,10 @@ NULL
         post.response=post.response,
         statement=statement,
         session.timezone=conn@session.timezone,
-        cursor=cursor
+        cursor=cursor,
+        update.conn.parameter=function(property, value) {
+          set_connection_parameter(conn, property, value)
```
Comment (onurfiliz) [inline R/dbSendQuery.R]: The RCpp route feels rather hacky. My suggestion would be to use a reference class (e.g. named PrestoSession) for storing the parameters in the connection - similar to what we are doing with the PrestoCursor. Then we can carry the instance of that class around in the PrestoResult to update as necessary.
We could have a method in PrestoConnection that returns the required list by reading the values from the reference class instance.

File: RPresto.Rproj (hunk 1 above)
Comment (saurfang) [inline RPresto.Rproj]: done

File: tests/testthat/test-session.property.R (hunk 2 above)
Comment (saurfang) [inline tests/testthat/test-session.property.R]: done

File: tests/testthat/test-session.property.R (hunk 3 above)
Comment (saurfang) [inline tests/testthat/test-session.property.R]: done

**Comment by saurfang** [comment]
Thank you for the review. I addressed all your comments and migrated to use a Reference class instead. However, I had to drop `parameters` slot from PrestoConnection to avoid storing potential conflict parameters in two places. The `dbConnect` interface is kept the same so unless users have been access `conn@parameters` directly, this should be fine. PTAL.

File: R/PrestoSession.R [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+    setParameter=function(key, value) {
+      .parameters[[key]] <<- value
+    },
+    unsetParameter=function(key) {
+      .parameters[[key]] <<- NULL
+    },
+    parameters=function() {
+      return(.parameters)
+    },
+    parameterStr=function() {
```
Comment (onurfiliz) [inline R/PrestoSession.R]: nit: I'd call this `parameterString`.

**Comment by saurfang** [comment]
Sure thing. I also squashed the commits and reran the tests. It seems fine from my end.

FYI some tests are failing for me because `'src_sql' is not an exported object from 'namespace:dplyr'` in `dplyr` release candidate.

**Comment by onurfiliz** [comment]
Hmm, the `src_sql` errors should be fixed by #76, you may have needed to rebase? In any case, I'll merge as is and verify that the tests pass. Thanks again.

**Pull request merged**